### PR TITLE
ARROW-16617: [C++] Add support for multi-byte system error message on Windows

### PR DESCRIPTION
**Pull request by kou**

### CODE REVIEW

File: cpp/src/arrow/util/io_util.cc [hunk 1]
```diff
@@ -203,16 +203,26 @@ std::string ErrnoMessage(int errnum) { return std::strerror(errnum); }
 
 #if _WIN32
 std::string WinErrorMessage(int errnum) {
-  char buf[1024];
-  auto nchars = FormatMessageA(FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_IGNORE_INSERTS,
-                               NULL, errnum, 0, buf, sizeof(buf), NULL);
-  if (nchars == 0) {
+#define MAX_N_CHARACTERS 1024
+#define MAX_N_UTF8_BYTES 4096
```
Comment (pitrou) [inline cpp/src/arrow/util/io_util.cc:207]: Can we use `constexpr` instead?

File: cpp/src/arrow/util/io_util.cc [hunk 2]
```diff
@@ -203,16 +203,26 @@ std::string ErrnoMessage(int errnum) { return std::strerror(errnum); }
...
     // Fallback
     std::stringstream ss;
     ss << "Windows error #" << errnum;
     return ss.str();
   }
-  return std::string(buf, nchars);
+  const UINT code_page = CP_UTF8;
+  const DWORD flags = 0;
+  char utf8_message[MAX_N_UTF8_BYTES];
+  auto n_utf8_bytes = WideCharToMultiByte(code_page, flags, utf16_message, n_utf16_chars,
```
Comment (pitrou) [inline cpp/src/arrow/util/io_util.cc:221]: You can use `WideStringToUTF8`.

**Review by pitrou — COMMENTED** [review]
Thanks. Did you validate that it fixes the issue?

File: cpp/src/arrow/util/io_util.cc (hunk 2 above)
Comment (pitrou) [reply cpp/src/arrow/util/io_util.cc:221]: Note that `WideCharToMultiByte` is converting to the current code page, but we want to convert to utf8 instead.

File: cpp/src/arrow/util/io_util.cc (hunk 1 above)
Comment (kou) [reply cpp/src/arrow/util/io_util.cc]: Yes. I should have used `constexpr`.

File: cpp/src/arrow/util/io_util.cc (hunk 2 above)
Comment (kou) [reply cpp/src/arrow/util/io_util.cc:221]: Who does provide `WideStringToUTF8`?
It seems that Windows doesn't provide it.

The code uses `CP_UTF8` not `CP_ACP` here. So this converts to UTF-8 not the current code page.

**Comment by kou** [comment]
> Did you validate that it fixes the issue?

Partially, yes. I just confirmed that C++ uses UTF-8 instead of not the current code page but didn't confirm for PyArrow.

I used `$BUILD_DIR/DEBUG/arrow-file-to-stream.exe XXX` to confirm this. It outputs `Could not convert to stream: IOError: Failed to open local file 'xxx'. Detail: [Windows error 2] 指定されたファイルが見つかりません。` on Japanese Windows. The output encoding was changed to UTF-8 from ShiftJIS (the current code page on Japanese Windows) with this change.

File: cpp/src/arrow/util/io_util.cc (hunk 2 above)
Comment (pitrou) [reply cpp/src/arrow/util/io_util.cc:221]: `WideStringToUTF8` is one of our helper functions.

Comment (kou) [reply cpp/src/arrow/util/io_util.cc:221]: Ah, thank. I could find it.

Comment (kou) [reply cpp/src/arrow/util/io_util.cc]: I've changed to use `WideStringToUTF8()`.

File: cpp/src/arrow/util/io_util.cc [hunk 3]
```diff
@@ -203,16 +203,25 @@ std::string ErrnoMessage(int errnum) { return std::strerror(errnum); }
...
     std::stringstream ss;
     ss << "Windows error #" << errnum;
     return ss.str();
   }
-  return std::string(buf, nchars);
+  auto utf8_message_result =
+      arrow::util::WideStringToUTF8(std::wstring(utf16_message, n_utf16_chars));
+  if (!utf8_message_result.ok()) {
+    std::stringstream ss;
+    ss << "Failed to convert to UTF-8: " << utf8_message_result.status();
```
Comment (pitrou) [inline cpp/src/arrow/util/io_util.cc:221]: Can we at least keep the error number here?
```suggestion
    ss << "Windows error #" << errnum;
    ss << "; failed to convert error message to UTF-8: " << utf8_message_result.status();
```

Comment (kou) [reply cpp/src/arrow/util/io_util.cc]: Yes.
It's a good idea.

**Review by pitrou — APPROVED** [review]
+1, let's wait for CI

**Comment by pitrou** [comment]
Now rebased.

**Pull request closed without merging**